### PR TITLE
TF initialization align with Torch

### PR DESCRIPTION
**Pull request by negvet**
- The following statistic collectors are implemented: `min_max`, `mean_min_max`, `threesigma`, `percentile`, `mean_percentile`.
- Per-layer range initialization.

### CODE REVIEW

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+import numpy as np
+import tensorflow as tf
+
+from nncf.tensorflow.layers.operation import InputType
+from nncf.tensorflow.quantization.initializers.utils import get_per_channel_history
+from nncf.tensorflow.quantization.initializers.utils import discard_zeros
+
+
+class MinMaxStatisticsCollector:
```
Comment (vshampor) [inline nncf/tensorflow/quantization/initializers/collectors.py:22]: I wonder if we can reduce the amount of copy-and-paste by using generics to reuse some of the logic between PT and TF stat collectors. Something to the tune of:
```python
class MinMaxStatisticsCollector(Generic[T])
   @abstractmethod
   def min(self) -> T:
          pass   # just as an example

class TFMinMaxStatisticsCollector(MinMaxStatisticsCollector[tf.Tensor])
    def min(self) -> tf.Tensor:
        # implement
```

Comment (vshampor) [reply nncf/tensorflow/quantization/initializers/collectors.py:22]: That, if successful, could also allow us to reuse tests. BTW, the unit tests are missing here.

Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py:22]: Thanks for the comment!

I agree. We have a plan to do so. For now, I provide only functional support and intentionally avoid any generics to better estimate the potential common part for TF and PT stat collectors'.

I am working on the unit tests.

**Comment by negvet** [comment]
Added tests

**Comment by negvet** [comment]
TODO:

- [ ] Add per layer range initialization

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

File: nncf/tensorflow/quantization/algorithm.py [hunk 2]
```diff
@@ -194,7 +183,17 @@ def get_transformation_layout(self, model: tf.keras.Model) -> TFTransformationLa
...
+    def _parse_init_params(self):
+        self._batchnorm_adaptation = 'batchnorm_adaptation' in self.config.get('initializer', {})
+        self._range_init_params = self._parse_range_init_params()
+
+    def _parse_range_init_params(self) -> TFRangeInitParams:
+        range_init_params = extract_range_init_params(self.config)
+        return TFRangeInitParams(**range_init_params) if range_init_params is not None else None
+
     def initialize(self, model: tf.keras.Model) -> None:
+        self._parse_init_params()
```
Comment (alexsu52) [inline nncf/tensorflow/quantization/algorithm.py:195]: You moved parsing parameters into initialize function. If initialization config is not correct than the user will see error relative to incorrect config only on initialization step. BTW this is the last step of the quantization pipeline. I would recomend to validate config into __init__ method.

**Review by alexsu52 — COMMENTED** [review]
It looks good to me.

File: nncf/tensorflow/quantization/algorithm.py (hunk 2 above)
Comment (negvet) [reply nncf/tensorflow/quantization/algorithm.py]: I agree, thanks. Fixed.

**Review by alexsu52 — APPROVED** [review]

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 1 above)
Comment (alexsu52) [reply nncf/tensorflow/quantization/initializers/collectors.py]: @nevget, Please open the ticket on @vshampor's comment. @vshampor Could you resolve conversation?

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
+"""
+ Copyright (c) 2020 Intel Corporation
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:2]:
```suggestion
 Copyright (c) 2021 Intel Corporation
```

**Review by vshampor — APPROVED** [review]

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+    @property
+    def max(self) -> tf.Tensor:
+        return tf.math.reduce_max(tf.stack(self.all_max_values), axis=0)
+
+    def prepare_statistics(self):
+        if self.per_channel:
+            new_shape = 1
+            for val in self.all_min_values[0].shape:
+                new_shape *= val
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:47]: Could we use `new_shape = np.prod(self.all_min_values[0].shape).item()` ?

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+from nncf.tensorflow.quantization.initializers.utils import get_axes
+
+
+class MinMaxStatisticCollector:
+    """
+    Collector uses min of minimum values and max of maximum values.
+    """
+
+    def __init__(self, per_channel: bool, channel_axes: int, input_type: str):
+        self.per_channel = per_channel
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:29]: Could you please use `_` prefix for all inner attributes?

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        self.call(*args, **kwargs)
+
+
+class MeanMinMaxStatisticsCollector:
+    """
+    Collector uses mean of minimum values and mean of maximum values.
+    """
+
+    def __init__(self, per_channel: bool, channel_axes: int, input_type: str):
+        self.per_channel = per_channel
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:76]: Could you please use `_` prefix for all inner attributes?

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        matches = []  # type: List[RangeInitConfig]
+        for pl_config in self.per_layer_range_init_configs:
+            if should_consider_scope(node_name,
+                                     ignored_scopes=pl_config.ignored_scopes,
+                                     target_scopes=pl_config.target_scopes):
+                if group == pl_config.target_group or pl_config.target_group is None:
+                    matches.append(RangeInitConfig(pl_config.init_type, pl_config.num_init_samples,
+                                                   pl_config.init_type_specific_params))
+        if len(matches) > 1:
+            raise ValueError("Location {} matches more than one per-layer initialization parameter "
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:68]: Could you please use `'` instead of `"`?

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                             "definition!".format(str(node_name)))
+        if len(matches) == 1:
+            return matches[0]
+        if not matches and self.global_init_config is not None:
+            return deepcopy(self.global_init_config)
+
+        raise ValueError("Location {} does not match any per-layer initialization parameter "
+                         "definition!".format(str(node_name)))
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:76]:
```suggestion
        raise ValueError(f'Location {node_name} does not match any per-layer initialization parameter '
                         'definition!')
```

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        range_type = init_config.init_type
+        if range_type == 'min_max':
+            return MinMaxStatisticCollector(per_channel, channel_axes, input_type)
+        if range_type == 'mean_min_max':
+            return MeanMinMaxStatisticsCollector(per_channel, channel_axes, input_type)
+        if range_type == 'threesigma':
+            return MedianMADStatisticCollector(per_channel, channel_axes, input_type)
+        if range_type == 'percentile':
+            min_percentile = init_config.init_type_specific_params.get("min_percentile", 0.1)
+            max_percentile = init_config.init_type_specific_params.get("max_percentile", 99.9)
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:98]:
```suggestion
            min_percentile = init_config.init_type_specific_params.get('min_percentile', 0.1)
            max_percentile = init_config.init_type_specific_params.get('max_percentile', 99.9)
```

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            return PercentileStatisticCollector(per_channel, channel_axes, input_type,
+                                                min_percentile, max_percentile)
+        if range_type == 'mean_percentile':
+            min_percentile = init_config.init_type_specific_params.get("min_percentile", 0.1)
+            max_percentile = init_config.init_type_specific_params.get("max_percentile", 99.9)
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:103]:
```suggestion
            min_percentile = init_config.init_type_specific_params.get('min_percentile', 0.1)
            max_percentile = init_config.init_type_specific_params.get('max_percentile', 99.9)
```

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+            return MeanPercentileStatisticCollector(per_channel, channel_axes, input_type,
+                                                    min_percentile, max_percentile)
+        raise ValueError('Range type {} is not supported.'.format(range_type))
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:106]:
```suggestion
        raise ValueError(f'Range type {range_type} is not supported.')
```

File: nncf/tensorflow/quantization/initializers/init_range.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    def run(self, model: tf.keras.Model) -> None:
+        layer_statistics = []
+        op_statistics = []
+        handles = []
+        for layer in model.layers:
+            if isinstance(layer, FakeQuantize):
+                channel_axes = get_channel_axis(InputType.INPUTS, '', layer)
+                init_config = self.range_init_params.get_init_config_for_quantization_point(layer, InputType.INPUTS)
+                collector = RangeInitializer.generate_stat_collector(init_config, layer.per_channel,
+                                                         channel_axes, InputType.INPUTS)
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/init_range.py:117]: Could you please align this?

File: nncf/tensorflow/quantization/initializers/utils.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+def get_per_channel_history(inputs_tensor: np.ndarray, axis: list) -> np.ndarray:
+    new_shape = 1
+    inputs_tensor_shape = inputs_tensor.shape
+    for dim in axis:
+        new_shape *= inputs_tensor_shape[dim]
+    return np.transpose(np.reshape(inputs_tensor, (new_shape, -1)))
+
+
+def get_axes(ndims, per_channel, channel_axes, add_dim=False):
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/utils.py:32]: Which name is correct for this method: `get_axes()` or `get_axis()`?

**Comment by negvet** [comment]
Jenkins please retry a build

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: nncf/tensorflow/quantization/initializers/utils.py (hunk 13 above)
Comment (vshampor) [reply nncf/tensorflow/quantization/initializers/utils.py:32]: https://www.merriam-webster.com/dictionary/axis

Looks like multiple instances of "axis" are being returned here, so "axes" is right.

Comment (andrey-churkin) [reply nncf/tensorflow/quantization/initializers/utils.py:32]: Ok, let's then rename the inner variable in this method from `axis` to `axes`.

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 3 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 4 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Good suggestion. Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 5 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 6 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 7 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 8 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 9 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 10 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 11 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/init_range.py (hunk 12 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/init_range.py]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/utils.py (hunk 13 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/utils.py:32]: Thanks, fixed!

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    @property
+    def max(self) -> tf.Tensor:
+        return tf.math.reduce_max(tf.stack(self._all_max_values), axis=0)
+
+    def prepare_statistics(self):
+        if self._per_channel:
+            new_shape = np.prod(self._all_min_values[0].shape).item()
+            for i, _ in enumerate(self._all_min_values):
+                self._all_min_values[i] = tf.reshape(self._all_min_values[i], shape=(new_shape))
+                self._all_max_values[i] = tf.reshape(self._all_max_values[i], shape=(new_shape))
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:48]:
```suggestion
                self._all_min_values[i] = tf.reshape(self._all_min_values[i], shape=new_shape)
                self._all_max_values[i] = tf.reshape(self._all_max_values[i], shape=new_shape)
```

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+    @property
+    def max(self) -> tf.Tensor:
+        return tf.math.reduce_mean(tf.stack(self._all_max_values), axis=0)
+
+    def prepare_statistics(self):
+        if self._per_channel:
+            new_shape = np.prod(self._all_min_values[0].shape).item()
+            for i, _ in enumerate(self._all_min_values):
+                self._all_min_values[i] = tf.reshape(self._all_min_values[i], shape=(new_shape))
+                self._all_max_values[i] = tf.reshape(self._all_max_values[i], shape=(new_shape))
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:93]:
```suggestion
                self._all_min_values[i] = tf.reshape(self._all_min_values[i], shape=new_shape)
                self._all_max_values[i] = tf.reshape(self._all_max_values[i], shape=new_shape)
```

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
```
Comment (andrey-churkin) [inline nncf/tensorflow/quantization/initializers/collectors.py:252]:
```suggestion
                self._all_min_values[i] = tf.reshape(self._all_min_values[i], shape=new_shape)
                self._all_max_values[i] = tf.reshape(self._all_max_values[i], shape=new_shape)
```

**Review by andrey-churkin — CHANGES_REQUESTED** [review]

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+from nncf.tensorflow.quantization.initializers.utils import discard_zeros
+from nncf.tensorflow.quantization.initializers.utils import get_axes
+
+
+class MinMaxStatisticCollector:
+    """
+    Collector uses min of minimum values and max of maximum values.
+    """
+
+    def __init__(self, per_channel: bool, channel_axes: int, input_type: str):
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+    def __call__(self, *args, **kwargs):
+        self.call(*args, **kwargs)
+
+
+class MedianMADStatisticCollector:
+    """
+    Collector uses three-sigma approach with the assumption of normal distribution by default.
+    """
+
+    def __init__(self, per_channel: bool, channel_axes: int, input_type: str):
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 15 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 16 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

File: nncf/tensorflow/quantization/initializers/collectors.py (hunk 14 above)
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

File: nncf/tensorflow/quantization/initializers/collectors.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+    def __call__(self, *args, **kwargs):
+        self.call(*args, **kwargs)
+
+
+class MeanMinMaxStatisticsCollector:
+    """
+    Collector uses mean of minimum values and mean of maximum values.
+    """
+
+    def __init__(self, per_channel: bool, channel_axes: int, input_type: str):
```
[the remark this answers is not in the document]
Comment (negvet) [reply nncf/tensorflow/quantization/initializers/collectors.py]: Thanks, done!

**Comment by negvet** [comment]
Jenkins please retry a build

**Comment by alexsu52** [comment]
Jenkins please retry a build

**Comment by alexsu52** [comment]
Jenkins please retry a build

**Review by andrey-churkin — APPROVED** [review]
